confidential-data-hub/hub/src/hub.rs

```diff
@@ -6,19 +6,79 @@
 use std::collections::HashMap;
 
 use async_trait::async_trait;
+use image_rs::image::ImageClient;
 use kms::{Annotations, ProviderSettings};
-use log::info;
+use log::{info, warn};
+use std::env;
 use storage::volume_type::Storage;
 
-use crate::{DataHub, Error, Result};
+use crate::{CdhConfig, DataHub, Error, Result};
 
 pub struct Hub {
     pub(crate) credentials: HashMap<String, String>,
+    image_client: tokio::sync::Mutex<image_rs::image::ImageClient>,
 }
 
 impl Hub {
-    pub async fn new(credentials: HashMap<String, String>) -> Result<Self> {
-        let mut hub = Self { credentials };
+    pub async fn new(config: CdhConfig) -> Result<Self> {
+        config.set_configuration_envs();
+        let credentials = config
+            .credentials
+            .iter()
+            .map(|it| (it.path.clone(), it.resource_uri.clone()))
+            .collect();
+
+        let mut image_client = match config.image.work_dir {
+            Some(work_dir) => ImageClient::new(work_dir.into()),
+            None => ImageClient::default(),
+        };
+
+        if let Some(uri) = config.image.extra_registry_auth_uri {
+            image_client.config.auth = true;
+            image_client.config.file_paths.auth_file = uri;
+        }
+
+        if let Some(uri) = config.image.image_security_policy_uri {
+            image_client.config.security_validate = true;
+            image_client.config.file_paths.policy_path = uri;
+        }
+
+        if let Some(uri) = config.image.sigstore_config_uri {
+            image_client.config.file_paths.sigstore_config = uri;
+        }
+
+        if let Some(number) = config.image.max_number_layer_download_currently {
+            image_client.config.max_concurrent_download = number;
+        }
+
+        // TODO: move the proxy envs to image-rs' PullClient once it supports
+        // Current the whole process of CDH would be influenced by the HTTPS_PROXY env
+        if let Some(https_proxy) = config.image.https_proxy {
+            match env::var("HTTPS_PROXY") {
```

Comment:
Are we trying to keep supporting the configs set by kata? I guess in the short-term that is a good idea, but in the long term I think the only kata configs that we should follow is `aa_kbc_params`.

Reply:
For now the only way to set the proxy of image pull is to set the env of the whole process. This would have some side effects where other parts also wants to set the proxy, because the overall proxy env would also influence the KBS client. The check here is to ensure no other env is set ahead of time.

There is a thread https://github.com/oras-project/rust-oci-client/pull/160 to make `oci-client` that image-rs is using to pull image to support the two proxy parameters only for the client struct. Once that gets merged, we can fix this.